source/common/grpc/google_grpc_utils.cc

```diff
@@ -0,0 +1,111 @@
+  std::atomic<uint32_t> ref_count_; // In case gPRC dereferences in a different threads.
+  Buffer::InstancePtr buffer_;
+
+  static void derefBufferInstanceContainer(void* container_ptr) {
+    auto container = reinterpret_cast<BufferInstanceContainer*>(container_ptr);
+    container->ref_count_--;
+    // This is safe because the ref_count_ is never incremented.
+    if (container->ref_count_ <= 0) {
+      delete container;
+    }
+  }
+};
+
+grpc::ByteBuffer GoogleGrpcUtils::makeByteBuffer(Buffer::InstancePtr&& buffer_instance) {
+  if (!buffer_instance) {
+    return {};
+  }
+  Buffer::RawSlice on_raw_slice;
+  // NB: we need to pass in >= 1 in order to get the real "n" (see Buffer::Instance for details).
+  int n_slices = buffer_instance->getRawSlices(&on_raw_slice, 1);
+  if (n_slices <= 0) {
+    return {};
+  }
+  auto container = new BufferInstanceContainer{n_slices, std::move(buffer_instance)};
+  if (n_slices == 1) {
+    grpc::Slice one_slice(on_raw_slice.mem_, on_raw_slice.len_,
+                          &BufferInstanceContainer::derefBufferInstanceContainer, container);
+    return {&one_slice, 1};
+  }
+  STACK_ARRAY(many_raw_slices, Buffer::RawSlice, n_slices);
+  container->buffer_->getRawSlices(many_raw_slices.begin(), n_slices);
+  std::vector<grpc::Slice> slices;
+  slices.reserve(n_slices);
+  for (int i = 0; i < n_slices; i++) {
+    slices.emplace_back(many_raw_slices[i].mem_, many_raw_slices[i].len_,
+                        &BufferInstanceContainer::derefBufferInstanceContainer, container);
+  }
+  return {&slices[0], slices.size()};
+}
+
+struct ByteBufferContainer {
+  ByteBufferContainer(int ref_count) : ref_count_(ref_count) {}
+  ~ByteBufferContainer() { ::free(fragments_); }
+  uint32_t ref_count_;
+  Buffer::BufferFragmentImpl* fragments_ = nullptr;
+  std::vector<grpc::Slice> slices_;
+};
+
+Buffer::InstancePtr GoogleGrpcUtils::makeBufferInstance(const grpc::ByteBuffer& byte_buffer) {
+  auto buffer = std::make_unique<Buffer::OwnedImpl>();
+  if (byte_buffer.Length() == 0) {
+    return buffer;
+  }
+  // NB: ByteBuffer::Dump moves the data out of the ByteBuffer so we need to ensure that the
+  // lifetime of the Slice(s) exceeds our Buffer::Instance.
+  std::vector<grpc::Slice> slices;
+  byte_buffer.Dump(&slices);
+  if (slices.empty()) {
+    return buffer;
+  }
+  auto* container = new ByteBufferContainer(static_cast<int>(slices.size()));
+  std::function<void(const void*, size_t, const Buffer::BufferFragmentImpl*)> releaser =
+      [container](const void*, size_t, const Buffer::BufferFragmentImpl*) {
+        container->ref_count_--;
+        if (container->ref_count_ <= 0) {
+          delete container;
+        }
+      };
+  // NB: addBufferFragment takes a pointer alias to the BufferFragmentImpl which is passed in so we
+  // need to ensure that the lifetime of those objects exceeds that of the Buffer::Instance.
+  RELEASE_ASSERT(!::posix_memalign(reinterpret_cast<void**>(&container->fragments_),
```

Comment:
Can you remind me why we have this really complicated memory alignment and manual memory management? Why can't `fragments_` be a vector of unique_ptr or something like that?

Reply:
lizan requested that I do this because of a problem with lua where a malloc of something which required non-standard alignment caused a problem.  That doesn't apply here but this is future-proof.